File: leetcode/BT_PostorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution2 {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        vector<int> ans;
        if (! root) return ans;
        
        stack<TreeNode *> s;
        s.push(root);
        
        while(! s.empty()) {
            TreeNode * n = s.top();
            s.pop();
            
            ans.push_back(n->val);
            
            if (n->left) s.push(n->left);
            if (n->right) s.push(n->right);
        }
        
        reverse(ans.begin(), ans.end());
        return ans;
    }
};


class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        
        stack<TreeNode * > s;
        TreeNode * n = root, * visited = root;
        
        while (n != NULL || ! s.empty()) {
            while (n != NULL) {
                s.push(n);
                n = n->left;
            }
            
            n = s.top();

            if (n->right == NULL || n->right == visited) {
                s.pop(); // do pop here!
                ans.push_back(n->val); // and here!

                visited = n;
                n = NULL;
            }
            else {
                n = n->right;
            }
        }
        
        return ans;
    }
};

File: leetcode_java/BT_PostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        if (root == null) return list;
        
        Stack<TreeNode> s = new Stack<TreeNode>();
        s.push(root);
        
        while (! s.empty()) {
            TreeNode n = s.peek();
            s.pop();
            list.add(n.val);
            
            if (n.left != null) s.push(n.left);
            if (n.right != null) s.push(n.right);
        }
        
        Collections.reverse(list);
        
        return list;
    }
}

/**
Binary Tree Postorder Traversal
Difficulty: Hard

Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},

   1
    \
     2
    /
   3

return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?
 */
